File: Embassat/View/Menu/CADEMMenuViewController.m
//
//  CADEMMenuViewController.m
//  Embassat
//
//  Created by Joan Romano on 22/03/14.
//  Copyright (c) 2014 Crows And Dogs. All rights reserved.
//

#import "CADEMMenuViewController.h"

#import "CADEMMenuCollectionViewCell.h"
#import "CADEMScheduleViewController.h"
#import "CADEMScheduleViewModel.h"

#import "UICollectionViewFlowLayout+EMAdditions.h"

@interface CADEMMenuViewController ()

@property (nonatomic, weak) IBOutlet UICollectionView *menuCollectionView;

@property (nonatomic, strong) CADArrayDataSourceSwift *dataSource;

@end

@implementation CADEMMenuViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    self.menuCollectionView.dataSource = self.dataSource;
    [self.menuCollectionView registerNib:[UINib nibWithNibName:@"CADEMMenuCollectionViewCell" bundle:nil] forCellWithReuseIdentifier:CADArrayDataSourceSwift.CADCellIdentifier];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    self.navigationController.navigationBarHidden = YES;
}

- (void)viewDidDisappear:(BOOL)animated
{
    [super viewDidDisappear:animated];
    
    for (NSIndexPath *indexPath in self.menuCollectionView.indexPathsForSelectedItems)
    {
        [self.menuCollectionView deselectItemAtIndexPath:indexPath animated:NO];
    }
}

#pragma mark - Collection View

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    id viewController;
    
    switch (indexPath.row)
    {
        case 0:
            viewController = [[CADEMInfoViewController alloc] initWithNibName:@"CADEMInfoViewController" bundle:nil];
            break;
            
        case 1:
            viewController = [[CADEMArtistsViewController alloc] initWithNibName:@"CADEMArtistsViewController" bundle:nil];
            break;
            
        case 2:
            viewController = [[CADEMScheduleViewController alloc] init];
            ((CADEMScheduleViewController *)viewController).viewModel = [[CADEMScheduleViewModel alloc] init];
            break;
            
        case 3:
            viewController = [[CADEMMapViewController alloc] initWithNibName:@"CADEMMapViewController" bundle:nil];
            break;
            
        case 4:
            viewController = [[CADEMTicketsViewController alloc] initWithNibName:@"CADEMTicketsViewController" bundle:nil];
            break;
            
        case 5:
            viewController = [[CADEMExtrasViewController alloc] initWithNibName:@"CADEMExtrasViewController" bundle:nil];
            break;
            
        default:
            break;
    }
    
    if (viewController)
    {
        [self.navigationController pushViewController:viewController animated:YES];
    }
}

#pragma mark - UICollectionViewFlowLayout Delegate

- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section
{
    UICollectionViewFlowLayout *flowLayout = (UICollectionViewFlowLayout *)collectionViewLayout;
    
    return [flowLayout insetsForVerticallyCenteredSectionInScreenWithRows:[self.dataSource.viewModel numberOfItemsInSection:0] andColumns:[self.dataSource.viewModel numberOfSections]];
}

#pragma mark - Lazy

- (CADArrayDataSourceSwift *)dataSource
{
    if (!_dataSource)
    {
        _dataSource = [[CADArrayDataSourceSwift alloc] initWithViewModel:self.viewModel configureCellBlock:^(CADEMMenuCollectionViewCell *cell, NSIndexPath *indexPath) {
            cell.optionName = [self.viewModel titleAtIndexPath:indexPath];
        } configureHeaderBlock:nil];
    }
    
    return _dataSource;
}

@end

File: Embassat/View/Menu/CADEMMenuViewController.swift
//
//  CADEMMenuViewController.swift
//  Embassa't
//
//  Created by Joan Romano on 25/05/15.
//  Copyright (c) 2015 Crows And Dogs. All rights reserved.
//

import Foundation

public class CADEMMenuViewController: CADEMRootViewControllerSwift {
    
    @IBOutlet weak var menuCollectionView: UICollectionView?
    let dataSource: CADArrayDataSourceSwift
    let viewModel: CADEMMenuViewModel
    
    override init(nibName nibNameOrNil: String?, bundle nibBundle: NSBundle?) {
        
        let theViewModel = CADEMMenuViewModel(model: ["Info", "Artistes", "Horaris", "Mapa", "Entrades", "Extres"])
        dataSource =
            CADArrayDataSourceSwift(viewModel: theViewModel,
                configureCellBlock: { (cell: AnyObject!, indexPath: NSIndexPath) -> Void in
                    let theCell = cell as! CADEMMenuCollectionViewCell
                    
                    theCell.optionName = theViewModel.titleAtIndexPath(indexPath)
                },
                configureHeaderBlock: nil)
        viewModel = theViewModel
        
        super.init(nibName: nibNameOrNil, bundle: nibBundle)
    }
    
    required public convenience init(coder aDecoder: NSCoder) {
        self.init(nibName: nil, bundle: nil)
    }

    public override func viewDidLoad() {
        super.viewDidLoad()

        menuCollectionView?.dataSource = dataSource
        menuCollectionView?.registerNib(UINib(nibName: "CADEMMenuCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: CADCellIdentifier)
    }
    
    public override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        self.navigationController?.navigationBarHidden = true
    }
    
    public override func viewDidDisappear(animated: Bool) {
        super.viewDidDisappear(animated)
        
        if let indexPaths = menuCollectionView?.indexPathsForSelectedItems() as? [NSIndexPath] {
            for indexPath in indexPaths {
                menuCollectionView?.deselectItemAtIndexPath(indexPath, animated: false)
            }
        }
    }
    
    func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
        var viewController: UIViewController?
        
        switch indexPath.item {
        case 0:
            viewController = CADEMInfoViewController(nibName: "CADEMInfoViewController", bundle: nil)
            break;
            
        case 1:
            viewController = CADEMArtistsViewController(nibName: "CADEMArtistsViewController", bundle: nil)
            break;
            
        case 2:
            viewController = CADEMScheduleViewController(nibName: "CADEMScheduleViewController", bundle: nil)
            break;
            
        case 3:
            viewController = CADEMMapViewController(nibName: "CADEMMapViewController", bundle: nil)
            break;
            
        case 4:
            viewController = CADEMTicketsViewController(nibName: "CADEMTicketsViewController", bundle: nil)
            break;
            
        case 5:
            viewController = CADEMExtrasViewController(nibName: "CADEMExtrasViewController", bundle: nil)
            break;
            
        default:
            break;
        }
        
        if viewController != nil {
            navigationController?.pushViewController(viewController!, animated: true)
        }
    }
    
    func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAtIndex section: Int) -> UIEdgeInsets {
        let flowLayout = collectionViewLayout as! UICollectionViewFlowLayout
        
        return flowLayout.insetsForVerticallyCenteredSectionInScreenWithRows(dataSource.viewModel.numberOfItemsInSection(0), andColumns: dataSource.viewModel.numberOfSections!())
    }
    
}